cmd/gvisor-website/main.go

```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
+
+	// Was this a dumb client? These are not supported.
+	vs, ok := r.URL.Query()["service"]
+	if !ok || len(r.URL.Query()) > 1 || len(vs) != 1 || vs[0] != "git-upload-pack" {
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprintf(w, "invalid query: %v", r.URL.Query())
+		return
+	}
+
+	// Connect upstream.
+	client := new(http.Client)
+	resp, err := client.Get(upstreamInfoRefs)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "upstream get error: %v", err)
+		return
+	}
+	defer resp.Body.Close()
+
+	// Read the full upstream contents.
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "upstream read error: %v", err)
+		return
+	}
+
+	// Was there an error? Pass all data through.
+	if resp.StatusCode != 200 {
+		w.WriteHeader(resp.StatusCode)
+		w.Write(data)
+		return
+	}
+
+	// emitPkt emits a single packet line.
+	emitPkt := func(m string) {
+		if len(m) == 0 {
+			// Special case: normally the size includes the size of
+			// the four byte header. However, we see that size 0000
+			// appears after the header and is the terminal. We use
+			// the empty string to indicate this.
+			fmt.Fprintf(w, "%04x", 0)
+		} else {
+			fmt.Fprintf(w, "%04x%s\n", 4+len(m)+1, m)
+		}
+	}
+
+	// readPkt reads a single packet line.
+	readPkt := func() (string, bool) {
+		// Parse the size header and return the string. Note that we
+		// don't do bounds checking here and rely on the recovery
+		// above. If we hit a bad reference, we will stop emitting
+		// references and return a bad response to the client.
+		if len(data) < 4 {
+			return "", false
+		}
+		size, err := strconv.ParseInt(string(data[:4]), 16, 32)
+		if err != nil {
+			return "", false
+		}
+		if size == 0 {
+			data = data[4:]
+			return "", true
+		} else if len(data) >= int(size) {
+			m := string(data[4:size])
+			data = data[size:]
+			return strings.TrimSuffix(m, "\n"), true
+		} else {
+			return "", false
+		}
+	}
+
+	const (
+		// serviceLine is the first line emitted.
+		serviceLine = "# service=git-upload-pack"
+
+		// target is the target branch.
+		target = "refs/heads/go"
+	)
+
+	// Check the header.
+	header, ok := readPkt()
+	if !ok || header != serviceLine {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "invalid upstream header: %v", header)
+		return
+	}
+
+	// readRef reads a single reference.
+	readRef := func() (hash string, ref string, options []string, ok bool) {
+		line, ok := readPkt()
+		if line == "" && ok {
+			return "", "", nil, true
+		}
+		if !ok || line == "" {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "invalid reference: %v", line)
+			return "", "", nil, false
+		}
+		parts := strings.Split(line, "\x00")
+		if len(parts) > 2 {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "invalid reference: %v", line)
+			return
+		} else if len(parts) == 2 {
```

Comment:
else isn't necessary

Reply:
If len(parts) is less than 2 then the following line will panic, no? (The len(parts) == 1 case isn't handled afaict). Unless of course readPkt guarantees that a string of len > 0 and without a delimiter is never returned somehow.

Reply:
The len(parts) == 1 case is handled below. Maybe you meant len(parts) == 0? That's guaranteed not to occur per the strings.Split semantics and the fact that I check for an empty string.

Reply:
Ah, ok, I misunderstood. The if is necessary but not the else.